File: conductor/hooks/subagent-contract.py
#!/usr/bin/env python3
"""SubagentStop: Verify t-plan subagent wrote required output.

This hook enforces t-plan "contracts" at the end of subagent runs:
- EXPLORE must write `explore.md`
- SCOUT must write `scout.md`
- VALIDATE must write `validation-vNNN.json` matching `draft_version`

The expected contract is determined by `.t-plan/<session_id>/state.json`.
"""

import json
import sys
from pathlib import Path
from typing import Optional

PHASE_CONTRACTS = {
    "EXPLORE": "explore.md",
    "SCOUT": "scout.md",
    "VALIDATE": "validation-v{version}.json",
}


def _find_session_dir(start: Path, session_id: str) -> Optional[Path]:
    # Walk upward so this works even if the hook runs from a subdirectory.
    for candidate_root in (start, *start.parents):
        session_dir = candidate_root / ".t-plan" / session_id
        if (session_dir / "state.json").exists():
            return session_dir
    return None


def _fail(message: str) -> None:
    print(message, file=sys.stderr)
    sys.exit(2)


def main():
    try:
        data = json.load(sys.stdin)
    except Exception:
        sys.exit(0)

    session_id = data.get("session_id")
    if not isinstance(session_id, str) or not session_id:
        sys.exit(0)

    cwd = Path(data.get("cwd") or Path.cwd())
    session_dir = _find_session_dir(cwd, session_id)
    if session_dir is None:
        sys.exit(0)

    state_file = session_dir / "state.json"
    try:
        state = json.loads(state_file.read_text())
    except Exception as exc:
        _fail(f"CONTRACT UNFULFILLED: invalid state.json ({exc})")

    phase = state.get("phase", "")

    if phase not in PHASE_CONTRACTS:
        sys.exit(0)

    expected = PHASE_CONTRACTS[phase]
    if "{version}" in expected:
        validation_version = state.get("validation_version")
        if isinstance(validation_version, str) and validation_version.isdigit():
            validation_version = int(validation_version)
        if not isinstance(validation_version, int) or validation_version < 1:
            _fail("CONTRACT UNFULFILLED: state.validation_version must be >= 1 for VALIDATE")
        expected = expected.format(version=str(validation_version).zfill(3))

    output = session_dir / expected

    if not output.exists() or output.stat().st_size == 0:
        _fail(f"CONTRACT UNFULFILLED: {phase} must write to {expected}")

    # Additional check for VALIDATE: JSON must be parseable and match draft_version
    if phase == "VALIDATE":
        try:
            validation = json.loads(output.read_text())
            file_draft = validation.get("draft_version")
            expected_draft = state.get("draft_version")

            try:
                file_draft_int = int(file_draft)
                expected_draft_int = int(expected_draft)
            except (TypeError, ValueError):
                _fail(
                    "CONTRACT UNFULFILLED: "
                    f"draft_version must be an integer (got {file_draft!r} vs {expected_draft!r})"
                )

            if file_draft_int != expected_draft_int:
                _fail(
                    "CONTRACT UNFULFILLED: "
                    f"validation draft_version ({file_draft}) != state ({expected_draft})"
                )
            if "status" not in validation:
                _fail("CONTRACT UNFULFILLED: validation JSON missing required field 'status'")
        except json.JSONDecodeError:
            _fail(f"CONTRACT UNFULFILLED: {expected} is not valid JSON")

    sys.exit(0)


if __name__ == "__main__":
    main()

File: conductor/hooks/subagent-contract.ts
#!/usr/bin/env bun
/**
 * Subagent Contract Hook (SubagentStop)
 *
 * Verifies that t-plan subagents fulfilled their contracts:
 * - EXPLORE must write explore.md
 * - SCOUT must write scout.md
 * - VALIDATE must write validation-vNNN.json with valid structure
 *
 * The expected contract is determined by .t-plan/{session_id}/state.json.
 */

import { existsSync, readFileSync, statSync } from "fs";
import { join } from "path";
import { readHookInput, allow, block } from "./lib/hooks";
import { findSessionDir, readState } from "./lib/state";
import { PHASE_CONTRACTS, type Phase } from "./lib/types";

function main(): never {
  const input = readHookInput();

  // Require session_id
  if (!input?.session_id) {
    return allow();
  }

  // Try to find session directory from various path candidates
  // Check multiple possible path fields (cwd is in HookInput, others may be added by Claude Code)
  const pathCandidates: string[] = [];
  const inputRecord = input as unknown as Record<string, unknown>;
  for (const key of ["cwd", "project_root", "repo_root", "workspace_root"]) {
    const value = inputRecord[key];
    if (typeof value === "string" && value) {
      pathCandidates.push(value);
    }
  }
  pathCandidates.push(process.cwd());

  let sessionDir: string | null = null;
  for (const candidate of pathCandidates) {
    sessionDir = findSessionDir(candidate, input.session_id);
    if (sessionDir) break;
  }

  // No t-plan session found - not a t-plan subagent
  if (!sessionDir) {
    return allow();
  }

  // Read state
  const state = readState(sessionDir);
  if (!state) {
    return allow();
  }

  const { phase, draft_version, validation_version } = state;

  // Unknown phase - allow
  if (!(phase in PHASE_CONTRACTS)) {
    return allow();
  }

  // Determine expected contract output
  let expected = PHASE_CONTRACTS[phase as Phase];
  if (expected.includes("{version}")) {
    if (typeof validation_version !== "number" || validation_version < 1) {
      return block(
        "CONTRACT UNFULFILLED: state.validation_version must be >= 1 for VALIDATE"
      );
    }
    expected = expected.replace(
      "{version}",
      String(validation_version).padStart(3, "0")
    );
  }

  const contractPath = join(sessionDir, expected);

  // Check file exists and is non-empty
  if (!existsSync(contractPath)) {
    return block(`CONTRACT UNFULFILLED: ${phase} must write ${expected}`);
  }

  const stats = statSync(contractPath);
  if (stats.size === 0) {
    return block(`CONTRACT UNFULFILLED: ${expected} is empty`);
  }

  // VALIDATE phase: additional JSON validation
  if (phase === "VALIDATE") {
    let validation: Record<string, unknown>;

    try {
      const content = readFileSync(contractPath, "utf8");
      validation = JSON.parse(content);
    } catch {
      return block(`CONTRACT UNFULFILLED: ${expected} is not valid JSON`);
    }

    // Check draft_version match
    const fileDraftVersion = validation.draft_version;
    const fileDraft =
      typeof fileDraftVersion === "number"
        ? fileDraftVersion
        : parseInt(String(fileDraftVersion), 10);
    const expectedDraft =
      typeof draft_version === "number"
        ? draft_version
        : parseInt(String(draft_version), 10);

    // parseInt returns NaN for invalid input (never throws)
    if (isNaN(fileDraft) || isNaN(expectedDraft)) {
      return block(
        `CONTRACT UNFULFILLED: draft_version must be an integer ` +
          `(got ${fileDraftVersion} vs ${draft_version})`
      );
    }

    if (fileDraft !== expectedDraft) {
      return block(
        `CONTRACT UNFULFILLED: validation draft_version (${fileDraft}) != state (${expectedDraft})`
      );
    }

    // Check required 'status' field
    if (!("status" in validation)) {
      return block(
        "CONTRACT UNFULFILLED: validation JSON missing required field 'status'"
      );
    }
  }

  return allow();
}

main();
